core/embed/rust/src/ui/shape/render.rs

```diff
@@ -0,0 +1,252 @@
+{
+    /// Target canvas
+    canvas: &'a mut C,
+    /// Bump for cloning shapes
+    bump: &'alloc T,
+    /// List of rendered shapes
+    shapes: FixedVec<'alloc, ShapeHolder<'alloc>>,
+    /// Current viewport
+    viewport: Viewport,
+    // Default background color
+    bg_color: Option<Color>,
+    /// Drawing cache (decompression context, scratch-pad memory)
+    cache: &'a DrawingCache<'alloc>,
+}
+
+impl<'a, 'alloc, T, C> ProgressiveRenderer<'a, 'alloc, T, C>
+where
+    T: LocalAllocLeakExt<'alloc>,
+    C: BasicCanvas,
+{
+    /// Creates a new ProgressiveRenderer instance
+    pub fn new(
+        canvas: &'a mut C,
+        bg_color: Option<Color>,
+        cache: &'a DrawingCache<'alloc>,
+        bump: &'alloc T,
+        max_shapes: usize,
+    ) -> Self {
+        let viewport = canvas.viewport();
+        Self {
+            canvas,
+            bump,
+            shapes: unwrap!(bump.fixed_vec(max_shapes), "No shape memory"),
+            viewport,
+            bg_color,
+            cache,
+        }
+    }
+
+    /// Renders stored shapes onto the specified canvas
+    pub fn render(&mut self, lines: usize) {
+        let canvas_clip = self.canvas.viewport().clip;
+
+        if canvas_clip.is_empty() {
+            return;
+        }
+
+        let buff = &mut unwrap!(self.cache.render_buff(), "No render buffer");
+
+        let mut slice = unwrap!(
+            Rgb565Canvas::new(
+                Offset::new(canvas_clip.width(), lines as i16),
+                None,
+                Some(1),
+                &mut buff[..],
+            ),
+            "No render memory"
+        );
+
+        for y in (canvas_clip.y0..canvas_clip.y1).step_by(lines) {
+            // Calculate the coordinates of the slice we will draw into
+            let slice_r = Rect::new(
+                // slice_r is in absolute coordinates
+                Point::new(canvas_clip.x0, y),
+                Point::new(canvas_clip.x1, y + lines as i16),
+            );
+
+            // Clear the slice background
+            if let Some(color) = self.bg_color {
+                slice.set_viewport(Viewport::from_size(slice_r.size()));
+                slice.fill_background(color);
+            }
+
+            // Draw all shapes that overlaps the slice
+            for holder in self.shapes.iter_mut() {
+                let shape_viewport = holder.viewport.absolute_clip(slice_r);
+                let shape_bounds = holder.shape.bounds(self.cache);
+
+                // Is the shape overlapping the current slice?
+                if shape_viewport.contains(shape_bounds) {
+                    slice.set_viewport(shape_viewport.translate((-slice_r.top_left()).into()));
+                    holder.shape.draw(&mut slice, self.cache);
+
+                    if shape_bounds.y1 + shape_viewport.origin.y <= shape_viewport.clip.y1 {
+                        // The shape will never be drawn again
+                        holder.shape.cleanup(self.cache);
```

Comment:
The main reason for having `cleanup()` was for optimization. It allows for manually clearing cached items and optimizing caching. However, it is no longer used for this purpose. We may consider omitting this concept in the future. But let it live for a while, it might prove useful.

I also realized that there is no need to have `&DrawingCache` as an argument in `Shape::bounds()`. Therefore, I removed it. See a983183078d8bc4ec0ef240e1df44bed35ef5dc5. This leads to some simplification and also make your first note irrelevant.